src/Server/HandlerCollection.cs

```diff
@@ -71,12 +71,12 @@ private HandlerDescriptor GetDescriptor(string method, Type implementedType, IJs
             }
 
             var key = "default";
-            if (handler is IRegistration<TextDocumentRegistrationOptions>)
+            // This protects against the case where class implements many, possibly conflicting, interfaces.
+            if (registration != null &&
+                typeof(TextDocumentRegistrationOptions).GetTypeInfo().IsAssignableFrom(registration) &&
+                handler is IRegistration<TextDocumentRegistrationOptions> handlerRegistration)
             {
-                if (GetTextDocumentRegistrationOptionsMethod
-                    .MakeGenericMethod(registration)
-                    .Invoke(handler, new object[] { handler }) is TextDocumentRegistrationOptions options)
-                    key = options.DocumentSelector;
+                key = handlerRegistration.GetRegistrationOptions()?.DocumentSelector ?? key;
```

Comment:
this might cause some interesting issues with some of the other derived types... I'll have to add a few more test cases to make sure won't fall over.

Reply:
I added some additional tests to make sure... so far so good I think.